### PR TITLE
Iterate over core classes C++11 style

### PR DESCRIPTION
**Pull request by ajrhacker**
C++11 range-based for loops can now iterate over simple_list, tagged_list, core_options, device_t::subdevice_list, device_t::interface_list, render_primitive_list and all subclasses of the above, and much code has been refactored to use them. Most core classes that have these lists as members now have methods that return the lists themselves, replacing most of the methods that returned the object at an owned list's head. (A few have been retained due to their use in drivers or OSD.)

device_t now manages subdevice and interface lists through subclasses, but has given up the work of adding and removing subdevices to machine_config.

memory_manager has its tagged lists exposed, though the old rooted tag lookup methods have been removed (they were privatized already).

### CODE REVIEW

File: src/emu/cheat.h [hunk 1]
```diff
@@ -294,7 +295,7 @@ class cheat_manager
 	// getters
 	running_machine &machine() const { return m_machine; }
 	bool enabled() const { return !m_disabled; }
-	cheat_entry *first() const { return m_cheatlist.first(); }
+	const auto &entries() const { return m_cheatlist; }
```
Comment (mmicko) [inline src/emu/cheat.h]: not sure this will work on GCC 4.9 which is minimal requirement, and highest available on some platforms.

**Comment by mmicko** [comment]
Can you please point to to those few you have retained ?

have pointed to const auto as return parameter, yes that is C++14 but please check if it is working under GCC4.9

**Comment by ajrhacker** [comment]
GCC 4.9, according to its release notes, should support auto return type deduction, but I don't have it for testing, and other compilers might not support that C++14 feature. I've decided to edit them out, and will commit this update once it successfully compiles.

The few first_xxx accessors I decided to retain are debug_view::first_source (used in OSD modules), software_info::first_part (used in the UI) and render_manager::first_target (used in a few drivers); these are redundant with the first() getters that all the iterator-savvy lists also provide. I do foresee all uses of these being rewritten later, but am a bit worried about breaking some WIPs; perhaps some of these cases should really be examining the whole list and not just the first entry.

**Comment by mmicko** [comment]
Sorry for late reply confirmed working on 4.9

**Comment by mmicko** [comment]
you can revert last one (	b0e0331) since it works on 4.9  and that fix for  unused is already in

**Comment by mmicko** [comment]
never mind, better this way :)

**Pull request merged**